src/encoder.rs

```diff
@@ -265,18 +368,34 @@ impl<W: Write> Writer<W> {
 
     /// Writes the image data.
     pub fn write_image_data(&mut self, data: &[u8]) -> Result<()> {
-        const MAX_CHUNK_LEN: u32 = (1u32 << 31) - 1;
+        const MAX_IDAT_CHUNK_LEN: u32 = std::u32::MAX >> 1;
+        const MAX_fdAT_CHUNK_LEN: u32 = (std::u32::MAX >> 1) - 4;
 
         if self.info.color_type == ColorType::Indexed && self.info.palette.is_none() {
             return Err(EncodingError::Format(FormatErrorKind::NoPalette.into()));
         }
 
-        let bpp = self.info.bpp_in_prediction();
-        let in_len = self.info.raw_row_length() - 1;
-        let prev = vec![0; in_len];
-        let mut prev = prev.as_slice();
-        let mut current = vec![0; in_len];
-        let data_size = in_len * self.info.height as usize;
+        let max = match self.info.animation_control {
+            Some(actl) if self.sep_def_img => actl.num_frames + 1,
```

Comment:
This could also lead to an overflow right?

Also considering that the sequence number and the number of frames are both 32 bits unsigned integers shouldn't it be impossible to encode an animation with more than u32::MAX / 2 frames? There is always at least a pair of chunks (fcTL and fdAT) for each frame so for each frame the sequence number is incremented at least twice

Reply:
You're right, it could. If anything, simply catch it here by returning an error. However, having a look at the [reference patch for libpng](https://sourceforge.net/projects/libpng-apng/files/libpng16/1.6.37/libpng-1.6.37-apng.patch.gz/download)  it seems that it simply overflows by using C's unsigned integer arithmetic—so kind of like a TCP sequence number. That would also make sense.

Reply:
I think that returning an error would be better as the spcification states that the sequence numbers are unique and strictly sequencial (no gaps)

Reply:
Okay, then add a check and error here if you want. Kind of awkward that the supposed C implementation doesn't actually implement the specification and I'm slightly skeptical that the mitigation of chunk reording warrants such a restriction but let's just be as spec compliant as possible.

Reply:
I decided to use wrapping_add as with the C implementation as there isn't really a good place to put the check, if you think it matters it could be solved in another PR, but we should think first at the `StreamWriter` implementation I think.
For the number of frames written I used an u64 so there is no need to check for overflows